File: sdk/java/src/main/java/io/openkedge/atp/internal/Manifest.java
package io.openkedge.atp.internal;

import static io.openkedge.atp.internal.Constants.D_MANIFEST;
import static io.openkedge.atp.internal.Constants.HARD_MAX_LEN;
import static io.openkedge.atp.internal.Constants.MAX_MANIFEST_BYTES;
import static io.openkedge.atp.internal.Constants.MAX_SCHEMA_ENUMS;
import static io.openkedge.atp.internal.Constants.MAX_SCHEMA_FIELDS;
import static io.openkedge.atp.internal.Constants.T_BYTES;
import static io.openkedge.atp.internal.Constants.T_ENUM;
import static io.openkedge.atp.internal.Constants.T_STRING;

import java.io.ByteArrayOutputStream;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * A schema manifest and its content-addressed identity. ATP-0002.
 *
 * <p>{@code H_S = SHA-256(D_MANIFEST || CanonicalCBOR(manifest))}. The manifest
 * bytes are canonical; the Java {@code record} is only a binding to them, which
 * is why a Java-derived {@code H_S} equals a Go/Rust/Python sibling's.
 */
public final class Manifest {
    public final String name;
    public final String version;
    public final int primitive;
    public final String publisher;
    public final List<FieldDef> fields;
    /** name -> members (ordinal order). Insertion order is irrelevant: the CBOR map sorts keys. */
    public final LinkedHashMap<String, List<String>> enums;
    /** optional (compatibilityVersion, mode); null if absent. */
    public final Object[] compatibility;

    public Manifest(String name, String version, int primitive, String publisher,
                    List<FieldDef> fields, LinkedHashMap<String, List<String>> enums,
                    Object[] compatibility) {
        this.name = name;
        this.version = version;
        this.primitive = primitive;
        this.publisher = publisher;
        this.fields = fields;
        this.enums = enums;
        this.compatibility = compatibility;
    }

    Cbor toCbor() {
        List<Map.Entry<Cbor, Cbor>> m = new ArrayList<>();
        m.add(Map.entry(Cbor.u(1), Cbor.text(name)));
        m.add(Map.entry(Cbor.u(2), Cbor.text(version)));
        m.add(Map.entry(Cbor.u(3), Cbor.u(primitive)));
        m.add(Map.entry(Cbor.u(4), Cbor.text(publisher)));

        List<Cbor> fieldArr = new ArrayList<>();
        for (FieldDef f : fields) {
            fieldArr.add(f.toCbor());
        }
        m.add(Map.entry(Cbor.u(5), new Cbor.Array(fieldArr)));

        List<Map.Entry<Cbor, Cbor>> enumMap = new ArrayList<>();
        for (Map.Entry<String, List<String>> e : enums.entrySet()) {
            List<Cbor> members = new ArrayList<>();
            for (String member : e.getValue()) {
                members.add(Cbor.text(member));
            }
            enumMap.add(Map.entry(Cbor.text(e.getKey()), new Cbor.Array(members)));
        }
        m.add(Map.entry(Cbor.u(6), new Cbor.MapC(enumMap)));

        if (compatibility != null) {
            String mcv = (String) compatibility[0];
            long mode = ((Number) compatibility[1]).longValue();
            List<Map.Entry<Cbor, Cbor>> cm = new ArrayList<>();
            cm.add(Map.entry(Cbor.u(1), Cbor.text(mcv)));
            cm.add(Map.entry(Cbor.u(2), Cbor.u(mode)));
            m.add(Map.entry(Cbor.u(7), new Cbor.MapC(cm)));
        }
        return new Cbor.MapC(m);
    }

    public byte[] canonicalCbor() {
        return Cbor.encode(toCbor());
    }

    public byte[] digest() {
        ByteArrayOutputStream pre = new ByteArrayOutputStream();
        pre.writeBytes(D_MANIFEST);
        pre.writeBytes(canonicalCbor());
        return Crypto.sha256(pre.toByteArray());
    }

    public List<String> enumMembers(String enumName) {
        return enums.get(enumName);
    }

    /**
     * Validate the manifest (a subset of ATP-0002 §5 sufficient for bind-time
     * fail-closed). Throws {@link IllegalArgumentException} with a stable reason.
     */
    public void validate() {
        if (name.length() > 128 || !validDottedName(name)) {
            throw err("BAD_SCHEMA_NAME");
        }
        if (!validSemver(version)) {
            throw err("BAD_SCHEMA_VERSION");
        }
        if (primitive > 3 || primitive < 0) {
            throw err("BAD_PRIMITIVE");
        }
        if (publisher.isEmpty() || publisher.length() > 128 || !allVchar(publisher)) {
            throw err("BAD_PUBLISHER");
        }
        if (fields.isEmpty() || fields.size() > MAX_SCHEMA_FIELDS) {
            throw err("EMPTY_FIELDS");
        }
        java.util.HashSet<String> fieldNames = new java.util.HashSet<>();
        for (int index = 0; index < fields.size(); index++) {
            FieldDef f = fields.get(index);
            if (f.slot != index) {
                throw err("SLOT_NOT_DENSE:expected=" + index + ":got=" + f.slot);
            }
            if (f.name.length() > 64 || !validFieldName(f.name)) {
                throw err("BAD_FIELD_NAME:" + index);
            }
            if (!fieldNames.add(f.name)) {
                throw err("DUP_FIELD_NAME:" + f.name);
            }
            if (f.type < 0 || f.type > 13) {
                throw err("UNKNOWN_TYPE:" + f.type);
            }
            if (f.unit != null && (f.unit.isEmpty() || f.unit.length() > 64 || !allVchar(f.unit))) {
                throw err("BAD_UNIT:" + f.name);
            }
            if (f.type == T_ENUM) {
                if (f.enumRef == null) {
                    throw err("ENUM_MISSING_ENUM_REF");
                }
                if (enumMembers(f.enumRef) == null) {
                    throw err("ENUM_REF_UNRESOLVED:" + f.enumRef);
                }
            } else if (f.enumRef != null) {
                throw err("ENUM_REF_ON_NONENUM:" + f.name);
            }
            if (f.name.equals("intent_ref")
                    && (!(primitive == 0 || primitive == 2) || f.type != T_BYTES)) {
                throw err("BAD_INTENT_REF");
            }
            validateConstraints(f);
        }
        if (enums.size() > MAX_SCHEMA_ENUMS) {
            throw err("TOO_MANY_ENUMS");
        }
        for (Map.Entry<String, List<String>> e : enums.entrySet()) {
            String enumName = e.getKey();
            List<String> members = e.getValue();
            if (enumName.length() > 64 || !validFieldName(enumName)) {
                throw err("BAD_ENUM_NAME");
            }
            if (members.isEmpty() || members.size() > 65_535) {
                throw err("EMPTY_ENUM:" + enumName);
            }
            java.util.HashSet<String> unique = new java.util.HashSet<>();
            for (String member : members) {
                if (member.isEmpty() || member.length() > 128 || !allVchar(member)) {
                    throw err("BAD_ENUM_MEMBER:" + enumName);
                }
                if (!unique.add(member)) {
                    throw err("DUP_ENUM_MEMBER:" + enumName);
                }
            }
        }
        if (canonicalCbor().length > MAX_MANIFEST_BYTES) {
            throw err("MANIFEST_TOO_LARGE");
        }
    }

    private void validateConstraints(FieldDef f) {
        java.util.HashSet<Long> keys = new java.util.HashSet<>();
        for (long[] kv : f.constraints) {
            if (!keys.add(kv[0])) {
                throw err("DUP_CONSTRAINT:" + f.name);
            }
            if (kv[0] == 1 && f.type != T_STRING && f.type != T_BYTES) {
                throw err("MAX_LEN_ON_NONLEN_TYPE:" + f.name);
            }
            if (kv[0] < 1 || kv[0] > 3) {
                throw err("UNKNOWN_CONSTRAINT:" + kv[0]);
            }
        }
        for (long[] kv : f.constraints) {
            if (kv[0] == 1 && (kv[1] < 1 || kv[1] > HARD_MAX_LEN)) {
                throw err("BAD_MAX_LEN:" + f.name);
            }
        }
    }

    private static IllegalArgumentException err(String reason) {
        return new IllegalArgumentException(reason);
    }

    private static boolean allVchar(String s) {
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c < 0x21 || c > 0x7e) {
                return false;
            }
        }
        return true;
    }

    static boolean validDottedName(String value) {
        if (value.isEmpty()) {
            return false;
        }
        char first = value.charAt(0);
        if (!(first >= 'a' && first <= 'z')) {
            return false;
        }
        boolean prevSep = false;
        for (int i = 1; i < value.length(); i++) {
            char c = value.charAt(i);
            if ((c >= 'a' && c <= 'z') || (c >= '0' && c <= '9')) {
                prevSep = false;
            } else if ((c == '.' || c == '_' || c == '-') && !prevSep) {
                prevSep = true;
            } else {
                return false;
            }
        }
        return !prevSep;
    }

    static boolean validFieldName(String value) {
        if (value.isEmpty()) {
            return false;
        }
        char first = value.charAt(0);
        if (!(first >= 'a' && first <= 'z')) {
            return false;
        }
        for (int i = 1; i < value.length(); i++) {
            char c = value.charAt(i);
            if (!((c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '_')) {
                return false;
            }
        }
        return true;
    }

    /** Strict SemVer core (major.minor.patch) with optional pre-release/build. */
    static boolean validSemver(String value) {
        if (value.isEmpty() || value.length() > 64) {
            return false;
        }
        for (int i = 0; i < value.length(); i++) {
            if (value.charAt(i) > 0x7f) {
                return false;
            }
        }
        String withoutBuild = value;
        int plus = value.indexOf('+');
        if (plus >= 0) {
            String build = value.substring(plus + 1);
            withoutBuild = value.substring(0, plus);
            if (!validSemverIdentifiers(build, false)) {
                return false;
            }
        }
        String core = withoutBuild;
        int dash = withoutBuild.indexOf('-');
        if (dash >= 0) {
            String pre = withoutBuild.substring(dash + 1);
            core = withoutBuild.substring(0, dash);
            if (!validSemverIdentifiers(pre, true)) {
                return false;
            }
        }
        String[] parts = core.split("\\.", -1);
        if (parts.length != 3) {
            return false;
        }
        for (String part : parts) {
            if (part.isEmpty() || !allDigits(part) || (!part.equals("0") && part.startsWith("0"))) {
                return false;
            }
        }
        return true;
    }

    private static boolean validSemverIdentifiers(String value, boolean rejectNumericLeadingZero) {
        if (value.isEmpty()) {
            return false;
        }
        for (String part : value.split("\\.", -1)) {
            if (part.isEmpty()) {
                return false;
            }
            for (int i = 0; i < part.length(); i++) {
                char c = part.charAt(i);
                if (!((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
                        || (c >= '0' && c <= '9') || c == '-')) {
                    return false;
                }
            }
            if (rejectNumericLeadingZero && allDigits(part) && !part.equals("0") && part.startsWith("0")) {
                return false;
            }
        }
        return true;
    }

    private static boolean allDigits(String s) {
        if (s.isEmpty()) {
            return false;
        }
        for (int i = 0; i < s.length(); i++) {
            if (!Character.isDigit(s.charAt(i))) {
                return false;
            }
        }
        return true;
    }
}

File: reference/rust/src/manifest.rs
//! Schema manifests and content-addressed identity. ATP-0002.

use crate::cbor::{encode, Cbor};
use crate::constants::{
    D_MANIFEST, HARD_MAX_LEN, MAX_MANIFEST_BYTES, MAX_SCHEMA_ENUMS, MAX_SCHEMA_FIELDS, T_BYTES,
    T_DURATION_MS, T_ENUM, T_F32, T_F64, T_I32, T_I64, T_STRING, T_TIMESTAMP_MS, T_U32, T_U64,
};
use crate::crypto::sha256;
use std::cmp::Ordering;
use std::collections::HashSet;

#[derive(Clone)]
pub struct Field {
    pub slot: u64,
    pub name: String,
    pub typ: u64,
    pub required: bool,
    pub unit: Option<String>,
    pub enum_ref: Option<String>,
    /// (constraint-key, value): 1=max_len, 2=min, 3=max (ATP-0002 §4.2).
    pub constraints: Vec<(u64, i64)>,
}

impl Field {
    pub fn new(slot: u64, name: &str, typ: u64, required: bool) -> Self {
        Field {
            slot,
            name: name.to_string(),
            typ,
            required,
            unit: None,
            enum_ref: None,
            constraints: Vec::new(),
        }
    }
    pub fn unit(mut self, u: &str) -> Self {
        self.unit = Some(u.to_string());
        self
    }
    pub fn enum_ref(mut self, e: &str) -> Self {
        self.enum_ref = Some(e.to_string());
        self
    }
    pub fn constraint(mut self, k: u64, v: i64) -> Self {
        self.constraints.push((k, v));
        self
    }

    fn to_cbor(&self) -> Cbor {
        let mut m = vec![
            (Cbor::U(1), Cbor::U(self.slot)),
            (Cbor::U(2), Cbor::Text(self.name.clone())),
            (Cbor::U(3), Cbor::U(self.typ)),
            (Cbor::U(4), Cbor::Bool(self.required)),
        ];
        if let Some(u) = &self.unit {
            m.push((Cbor::U(5), Cbor::Text(u.clone())));
        }
        if let Some(er) = &self.enum_ref {
            m.push((Cbor::U(6), Cbor::Text(er.clone())));
        }
        if !self.constraints.is_empty() {
            let cm = self
                .constraints
                .iter()
                .map(|(k, v)| (Cbor::U(*k), Cbor::I(*v)))
                .collect();
            m.push((Cbor::U(7), Cbor::Map(cm)));
        }
        Cbor::Map(m)
    }
}

#[derive(Clone)]
pub struct Manifest {
    pub name: String,
    pub version: String,
    pub primitive: u64,
    pub publisher: String,
    pub fields: Vec<Field>,
    pub enums: Vec<(String, Vec<String>)>,
    pub compatibility: Option<(String, u64)>,
}

impl Manifest {
    pub fn validate(&self) -> Result<(), String> {
        if self.name.len() > 128 || !valid_dotted_name(&self.name) {
            return Err("BAD_SCHEMA_NAME".into());
        }
        if !valid_semver(&self.version) {
            return Err("BAD_SCHEMA_VERSION".into());
        }
        if self.primitive > 3 {
            return Err("BAD_PRIMITIVE".into());
        }
        if self.publisher.is_empty()
            || self.publisher.len() > 128
            || !self.publisher.bytes().all(|b| (0x21..=0x7e).contains(&b))
        {
            return Err("BAD_PUBLISHER".into());
        }
        if self.fields.is_empty() || self.fields.len() > MAX_SCHEMA_FIELDS {
            return Err("EMPTY_FIELDS".into());
        }

        let mut field_names = HashSet::new();
        for (index, field) in self.fields.iter().enumerate() {
            if field.slot != index as u64 {
                return Err(format!(
                    "SLOT_NOT_DENSE:expected={index}:got={}",
                    field.slot
                ));
            }
            if field.name.len() > 64 || !valid_field_name(&field.name) {
                return Err(format!("BAD_FIELD_NAME:{index}"));
            }
            if !field_names.insert(field.name.as_str()) {
                return Err(format!("DUP_FIELD_NAME:{}", field.name));
            }
            if field.typ > 13 {
                return Err(format!("UNKNOWN_TYPE:{}", field.typ));
            }
            if let Some(unit) = &field.unit {
                if unit.is_empty()
                    || unit.len() > 64
                    || !unit.bytes().all(|b| (0x21..=0x7e).contains(&b))
                {
                    return Err(format!("BAD_UNIT:{}", field.name));
                }
            }
            if field.typ == T_ENUM {
                let enum_ref = field
                    .enum_ref
                    .as_ref()
                    .ok_or_else(|| "ENUM_MISSING_ENUM_REF".to_string())?;
                if self.enum_members(enum_ref).is_none() {
                    return Err(format!("ENUM_REF_UNRESOLVED:{enum_ref}"));
                }
            } else if field.enum_ref.is_some() {
                return Err(format!("ENUM_REF_ON_NONENUM:{}", field.name));
            }
            if field.name == "intent_ref"
                && (!(self.primitive == 0 || self.primitive == 2) || field.typ != T_BYTES)
            {
                return Err("BAD_INTENT_REF".into());
            }
            validate_constraints(field)?;
        }

        if self.enums.len() > MAX_SCHEMA_ENUMS {
            return Err("TOO_MANY_ENUMS".into());
        }
        let mut enum_names = HashSet::new();
        for (name, members) in &self.enums {
            if name.len() > 64 || !valid_field_name(name) {
                return Err("BAD_ENUM_NAME".into());
            }
            if !enum_names.insert(name.as_str()) {
                return Err(format!("DUP_ENUM:{name}"));
            }
            if members.is_empty() || members.len() > 65_535 {
                return Err(format!("EMPTY_ENUM:{name}"));
            }
            let mut unique = HashSet::new();
            for member in members {
                if member.is_empty()
                    || member.len() > 128
                    || !member.bytes().all(|b| (0x21..=0x7e).contains(&b))
                {
                    return Err(format!("BAD_ENUM_MEMBER:{name}"));
                }
                if !unique.insert(member.as_str()) {
                    return Err(format!("DUP_ENUM_MEMBER:{name}"));
                }
            }
        }

        if let Some((version, mode)) = &self.compatibility {
            if !valid_semver(version) {
                return Err("BAD_COMPATIBILITY_VERSION".into());
            }
            if *mode > 3 {
                return Err("BAD_COMPATIBILITY_MODE".into());
            }
            let ordering = compare_semver(version, &self.version)
                .ok_or_else(|| "BAD_COMPATIBILITY_VERSION".to_string())?;
            if ordering == Ordering::Greater || (*mode == 0 && ordering != Ordering::Equal) {
                return Err("BAD_COMPATIBILITY_RANGE".into());
            }
        }
        if self.canonical_cbor().len() > MAX_MANIFEST_BYTES {
            return Err("MANIFEST_TOO_LARGE".into());
        }
        Ok(())
    }

    pub fn to_cbor(&self) -> Cbor {
        let mut m = vec![
            (Cbor::U(1), Cbor::Text(self.name.clone())),
            (Cbor::U(2), Cbor::Text(self.version.clone())),
            (Cbor::U(3), Cbor::U(self.primitive)),
            (Cbor::U(4), Cbor::Text(self.publisher.clone())),
            (
                Cbor::U(5),
                Cbor::Array(self.fields.iter().map(|f| f.to_cbor()).collect()),
            ),
            (
                Cbor::U(6),
                Cbor::Map(
                    self.enums
                        .iter()
                        .map(|(k, vs)| {
                            (
                                Cbor::Text(k.clone()),
                                Cbor::Array(vs.iter().map(|x| Cbor::Text(x.clone())).collect()),
                            )
                        })
                        .collect(),
                ),
            ),
        ];
        if let Some((mcv, mode)) = &self.compatibility {
            m.push((
                Cbor::U(7),
                Cbor::Map(vec![
                    (Cbor::U(1), Cbor::Text(mcv.clone())),
                    (Cbor::U(2), Cbor::U(*mode)),
                ]),
            ));
        }
        Cbor::Map(m)
    }

    pub fn canonical_cbor(&self) -> Vec<u8> {
        encode(&self.to_cbor())
    }

    pub fn digest(&self) -> [u8; 32] {
        let mut pre = D_MANIFEST.to_vec();
        pre.extend(self.canonical_cbor());
        sha256(&pre)
    }

    pub fn enum_members(&self, name: &str) -> Option<&Vec<String>> {
        self.enums.iter().find(|(k, _)| k == name).map(|(_, v)| v)
    }
}

fn valid_dotted_name(value: &str) -> bool {
    let mut chars = value.chars();
    if !matches!(chars.next(), Some(c) if c.is_ascii_lowercase()) {
        return false;
    }
    let mut previous_separator = false;
    for c in chars {
        if c.is_ascii_lowercase() || c.is_ascii_digit() {
            previous_separator = false;
        } else if matches!(c, '.' | '_' | '-') && !previous_separator {
            previous_separator = true;
        } else {
            return false;
        }
    }
    !previous_separator
}

fn valid_field_name(value: &str) -> bool {
    let mut chars = value.chars();
    matches!(chars.next(), Some(c) if c.is_ascii_lowercase())
        && chars.all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_')
}

fn valid_semver(value: &str) -> bool {
    if value.is_empty() || value.len() > 64 || !value.is_ascii() {
        return false;
    }
    let (without_build, build) = value
        .split_once('+')
        .map_or((value, None), |(left, right)| (left, Some(right)));
    if build
        .map(|ids| !valid_semver_identifiers(ids, false))
        .unwrap_or(false)
    {
        return false;
    }
    let (core, prerelease) = without_build
        .split_once('-')
        .map_or((without_build, None), |(left, right)| (left, Some(right)));
    if prerelease
        .map(|ids| !valid_semver_identifiers(ids, true))
        .unwrap_or(false)
    {
        return false;
    }
    let mut parts = core.split('.');
    let (Some(major), Some(minor), Some(patch), None) =
        (parts.next(), parts.next(), parts.next(), parts.next())
    else {
        return false;
    };
    [major, minor, patch].iter().all(|part| {
        !part.is_empty()
            && part.bytes().all(|b| b.is_ascii_digit())
            && (*part == "0" || !part.starts_with('0'))
    })
}

fn valid_semver_identifiers(value: &str, reject_numeric_leading_zero: bool) -> bool {
    !value.is_empty()
        && value.split('.').all(|part| {
            !part.is_empty()
                && part.bytes().all(|b| b.is_ascii_alphanumeric() || b == b'-')
                && (!reject_numeric_leading_zero
                    || !part.bytes().all(|b| b.is_ascii_digit())
                    || part == "0"
                    || !part.starts_with('0'))
        })
}

fn compare_semver(left: &str, right: &str) -> Option<Ordering> {
    if !valid_semver(left) || !valid_semver(right) {
        return None;
    }
    let (left_core, left_pre) = split_semver(left);
    let (right_core, right_pre) = split_semver(right);
    for (a, b) in left_core.split('.').zip(right_core.split('.')) {
        let ordering = a.len().cmp(&b.len()).then_with(|| a.cmp(b));
        if ordering != Ordering::Equal {
            return Some(ordering);
        }
    }
    match (left_pre, right_pre) {
        (None, None) => Some(Ordering::Equal),
        (None, Some(_)) => Some(Ordering::Greater),
        (Some(_), None) => Some(Ordering::Less),
        (Some(a), Some(b)) => {
            let mut ai = a.split('.');
            let mut bi = b.split('.');
            loop {
                match (ai.next(), bi.next()) {
                    (None, None) => return Some(Ordering::Equal),
                    (None, Some(_)) => return Some(Ordering::Less),
                    (Some(_), None) => return Some(Ordering::Greater),
                    (Some(x), Some(y)) => {
                        let x_numeric = x.bytes().all(|byte| byte.is_ascii_digit());
                        let y_numeric = y.bytes().all(|byte| byte.is_ascii_digit());
                        let ordering = match (x_numeric, y_numeric) {
                            (true, true) => x.len().cmp(&y.len()).then_with(|| x.cmp(y)),
                            (true, false) => Ordering::Less,
                            (false, true) => Ordering::Greater,
                            (false, false) => x.cmp(y),
                        };
                        if ordering != Ordering::Equal {
                            return Some(ordering);
                        }
                    }
                }
            }
        }
    }
}

fn split_semver(value: &str) -> (&str, Option<&str>) {
    let without_build = value.split_once('+').map_or(value, |(head, _)| head);
    without_build
        .split_once('-')
        .map_or((without_build, None), |(core, pre)| (core, Some(pre)))
}

fn validate_constraints(field: &Field) -> Result<(), String> {
    let mut keys = HashSet::new();
    for (key, value) in &field.constraints {
        if !keys.insert(*key) {
            return Err(format!("DUP_CONSTRAINT:{}", field.name));
        }
        match *key {
            1 if field.typ != T_STRING && field.typ != T_BYTES => {
                return Err(format!("MAX_LEN_ON_NONLEN_TYPE:{}", type_name(field.typ)));
            }
            2 | 3
                if !matches!(
                    field.typ,
                    T_U32 | T_U64 | T_I32 | T_I64 | T_F32 | T_F64 | T_TIMESTAMP_MS | T_DURATION_MS
                ) =>
            {
                return Err(format!("RANGE_ON_NONNUMERIC:{}", type_name(field.typ)));
            }
            1..=3 => {
                let _ = value;
            }
            _ => return Err(format!("UNKNOWN_CONSTRAINT:{key}")),
        }
    }
    let get = |key| {
        field
            .constraints
            .iter()
            .find(|(k, _)| *k == key)
            .map(|x| x.1)
    };
    if let Some(max_len) = get(1) {
        if max_len < 1 || max_len as u64 > HARD_MAX_LEN {
            return Err(format!("BAD_MAX_LEN:{}", field.name));
        }
    }
    let (min, max) = (get(2), get(3));
    if min.zip(max).map(|(a, b)| a > b).unwrap_or(false) {
        return Err(format!("MIN_GT_MAX:{}", field.name));
    }
    if matches!(field.typ, T_U32 | T_U64 | T_DURATION_MS)
        && min.into_iter().chain(max).any(|v| v < 0)
    {
        return Err(format!("UNSIGNED_RANGE_OUT_OF_DOMAIN:{}", field.name));
    }
    if field.typ == T_U32
        && min
            .into_iter()
            .chain(max)
            .any(|v| v as u64 > u32::MAX as u64)
    {
        return Err(format!("UNSIGNED_RANGE_OUT_OF_DOMAIN:{}", field.name));
    }
    if field.typ == T_I32
        && min
            .into_iter()
            .chain(max)
            .any(|v| v < i32::MIN as i64 || v > i32::MAX as i64)
    {
        return Err(format!("I32_RANGE_OUT_OF_DOMAIN:{}", field.name));
    }
    Ok(())
}

fn type_name(typ: u64) -> &'static str {
    match typ {
        0 => "BOOL",
        1 => "U32",
        2 => "U64",
        3 => "I32",
        4 => "I64",
        5 => "F32",
        6 => "F64",
        7 => "ENUM",
        8 => "STRING",
        9 => "BYTES",
        10 => "TIMESTAMP_MS",
        11 => "ENTITY_REF",
        12 => "OPAQUE_REF",
        13 => "DURATION_MS",
        _ => "UNKNOWN",
    }
}

#[cfg(test)]
mod tests {
    use super::{compare_semver, valid_semver};
    use std::cmp::Ordering;

    #[test]
    fn semver_syntax_is_strict() {
        for valid in [
            "0.0.0",
            "1.2.3",
            "1.0.0-alpha.1",
            "1.0.0+build.7",
            "1.0.0-rc.1+sha.abc",
        ] {
            assert!(valid_semver(valid), "{valid}");
        }
        for invalid in [
            "", "v1.0.0", "1.0", "01.0.0", "1.0.0-", "1.0.0-01", "1.0.0+", "1.0.0++x",
        ] {
            assert!(!valid_semver(invalid), "{invalid}");
        }
    }

    #[test]
    fn semver_precedence_ignores_build_metadata() {
        assert_eq!(
            compare_semver("1.0.0-alpha.2", "1.0.0-alpha.10"),
            Some(Ordering::Less)
        );
        assert_eq!(compare_semver("1.0.0-alpha", "1.0.0"), Some(Ordering::Less));
        assert_eq!(
            compare_semver("1.0.0+left", "1.0.0+right"),
            Some(Ordering::Equal)
        );
    }
}
